### ISSUE TITLE
No pending data given in read callback after a connection closes and a write fails

### ISSUE THREAD

**Issue by Crypthead** [libuv/libuv#2643]
* **Version**:  Version 1.34.1 (Stable)
* **Platform**: Linux Hostname 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

Hi, I am using Libuv as a proxy between a client and a server where the server transfers data while the client sends feedback data, and all I do with Libuv is forward it between them. When the server is done it closes the connection even if feedback is being sent. At some point it seems like I reach a corner case with libuv, where both a read callback is set and a write callback is set, and then the following happens inconsistently (~1/10 transfers):

1. The write callback is called with an error code, I return without doing anything.
2. The read callback is called with an EOF, but when looking at the client not all data has been transferred (the last bytes are missing).

In successful transfers, there never are any failed write callbacks.

Investigating the issue, after the error in the write callback I grabbed the file descriptor from the handle and did a manual read which gave me said pending data. This leads me to believe the failed write callback causes libuv to call the read callback with an EOF even though it should be calling it once before with data that is still available. Maybe the data is discarded since the connection is broken at some point during the treatment of the failed write?

Is it possible that libuv withholds said data because the connection has broken? If not, just close the issue ignore everything I just said.

Some more details: testing done on localhost, server application serves data from memory so everything goes very fast.

**Comment by vtjnash** [comment]
Do you have some sample code? It doesn't seem like write failure should stop you from being able to drain the existing read buffer, but after some research, I'm beginning to wonder if that's the BSD sockets design. I can show something similar on macOS (x86_64-apple-darwin18.7.0), but I see the short read / error even if bypassing libuv:
```julia
julia> using Random, Sockets

julia> t = listen(0)
Sockets.TCPServer(RawFD(0x00000015) active)

julia> getsockname(t)
(ip"127.0.0.1", 0xd6a7)

julia> c = connect(ans[2])
TCPSocket(RawFD(0x00000016) open, 0 bytes waiting)

julia> r = accept(t)
TCPSocket(RawFD(0x00000017) open, 0 bytes waiting)

julia> close(t)

julia> @async (write(c, randstring(10_000_000)); close(c); @show "done write")
Task (runnable) @0x0000000117b94010

julia> tot = 0
0

julia> @async(write(r, "1")), while !eof(r); @show global tot += sizeof(readavailable(r)); end
global tot += sizeof(readavailable(r)) = 1455160
global tot += sizeof(readavailable(r)) = 2739328
global tot += sizeof(readavailable(r)) = 4269100
global tot += sizeof(readavailable(r)) = 5635188
global tot += sizeof(readavailable(r)) = 7181396
global tot += sizeof(readavailable(r)) = 8735796
"done write" = "done write"
global tot += sizeof(readavailable(r)) = 9834200
ERROR: IOError: read: connection reset by peer (ECONNRESET)
Stacktrace:
 [1] wait_readnb(::TCPSocket, ::Int64) at ./stream.jl:336
 [2] eof(::TCPSocket) at ./stream.jl:48
 [3] top-level scope at task.jl:367

julia> tot
9829774
```

or alternate reads going directly to the kernel:
```julia
julia> b = zeros(UInt8, 1_000_000); (tot += @show(ccall(:read, Csize_t, (Cint, Ptr{UInt8}, Csize_t), 0x17, b, sizeof(b)))), Libc.errno(), b
ccall(:read, Csize_t, (Cint, Ptr{UInt8}, Csize_t), 0x17, b, sizeof(b)) = 0x00000000000ebe2c
(0x000000000092df18, 0, UInt8[0x68, 0x70, 0x75, 0x4e, 0x6f, 0x5a, 0x47, 0x39, 0x71, 0x78  …  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])

julia> b = zeros(UInt8, 1_000_000); (tot += @show(ccall(:read, Csize_t, (Cint, Ptr{UInt8}, Csize_t), 0x17, b, sizeof(b)))), Libc.errno(), b
ccall(:read, Csize_t, (Cint, Ptr{UInt8}, Csize_t), 0x17, b, sizeof(b)) = 0xffffffffffffffff
(0x000000000092df17, 54, UInt8[0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  …  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])

julia> tot
0x0000000000989680
```

**Comment by Crypthead** [comment]
After some more investigation I stumbled upon [the following article](https://www.oreilly.com/library/view/http-the-definitive/1565925092/ch04s07.html) mainly about HTTP but the only relevant information there are the last 2 sections which talk about TCP. Tl;dr: a connection reset by peer can cause certain operating system to erase buffered, unread data. I'll try to confirm this with some traffic captures but this more than likely is my problem.

Closing since the issue isn't relevant to libuv, and more related to the server. Sorry for the waste of time :)
